### PR TITLE
.Net: Add support for ImageContent to use data URIs in ChatPromptParser so templates can use base64 encoded images.

### LINKED ISSUE #7150: .Net: Is it possible to use binary image data with a handlebars template?
### Discussed in https://github.com/microsoft/semantic-kernel/discussions/7121

<div type='discussions-op-text'>

<sup>Originally posted by **jmservera** July  6, 2024</sup>
I'm trying to do something like this:

```csharp
var getDescription= kernel.CreateFunctionFromPrompt(
    """
    <message role="system>You are an AI assistant that based on a picture or its description you transform it to a funny title
    in English, French and Spanish.
    The output should be a json file with "en", "fr" and "es" as keys for the translations.
    Here is the output schema:
        {
         "en":"English description",
         "fr": "French translation",
         "es": "Spanish translation"
        }
    </message>
    <message role="user">
    {{#if image}}
    <image>
    {{image}}
    </image>
    {{/if}}
    <text>
    {{#if caption}}
    Here are the captions for the picture: {{caption}}
    {{/if}}
    {{#if people}}
    Here's the people that appear in the picture: {{ people }}
    {{/if}}
    Please provide a funny sentence for this.
    </text>
    </message>
    """, templateFormat: "handlebars",promptTemplateFactory: new HandlebarsPromptTemplateFactory());
```
but the &lt;image&gt; tag seems to only take uri as valid value. I've tried with ImageContent's DataUri but then I get a token limit exception:

```csharp
ImageContent image=new(new ReadOnlyMemory<byte>(bytes), "image/jpeg");
    parameters.Add("image",image.DataUri);
```

Thanks,
Juanma</div>

### PR DESCRIPTION
**Pull request by MarceloAGG**
### Motivation and Context

At present, including images in prompt templates using base64 data encoding is not possible. This limitation is due to `ChatPromptParser.cs` exclusively calling the `ImageContent` constructor that requires a URI, which leads to an `InvalidOperationException`. The change required is straightforward and the limitation has been discussed before, [for example here](https://github.com/microsoft/semantic-kernel/discussions/7121). Closes #7150.

### Description

The proposed trivial fix involves a simple check to determine if the content starts with `data:`, and if it does the `ImageContent` constructor that accepts a `dataUri` is utilized instead. 

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by MarceloAGG** [comment]
@microsoft-github-policy-service agree company="Microsoft"

**Comment by MarceloAGG** [comment]
Good suggestion @RogerBarreto, `Concept` sample added.

File: dotnet/samples/Concepts/README.md [hunk 1]
```diff
@@ -142,7 +142,8 @@ Down below you can find the code snippets that demonstrate the usage of many Sem
 - [MultiplePromptTemplates](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/PromptTemplates/MultiplePromptTemplates.cs)
 - [PromptFunctionsWithChatGPT](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/PromptTemplates/PromptFunctionsWithChatGPT.cs)
 - [TemplateLanguage](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/PromptTemplates/TemplateLanguage.cs)
-- [PromptyFunction](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/PromptYemplates/PromptyFunction.cs)
+- [PromptyFunction](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/PromptTemplates/PromptyFunction.cs)
```
Comment (MarceloAGG) [inline dotnet/samples/Concepts/README.md:145]: There was a typo here: `PromptYemplates`.

File: dotnet/samples/Concepts/PromptTemplates/HandlebarsVisionPrompts.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        var function = kernel.CreateFunctionFromPrompt(promptTemplateConfig, templateFactory);
+
+        var arguments = new KernelArguments(new Dictionary<string, object?>
+        {
+            {"request","Describe this image:"},
+            {"imageData", "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAAXNSR0IArs4c6QAAACVJREFUKFNj/KTO/J+BCMA4iBUyQX1A0I10VAizCj1oMdyISyEAFoQbHwTcuS8AAAAASUVORK5CYII="}
+        });
+
+        var response = await kernel.InvokeAsync(function, arguments);
+        Console.WriteLine(response.GetValue<string?>());
```
Comment (RogerBarreto) [inline dotnet/samples/Concepts/PromptTemplates/HandlebarsVisionPrompts.cs:44]: Response already has an override that tries to get value as string.

```suggestion
        Console.WriteLine(response);
```

**Review by RogerBarreto — APPROVED** [review]

**Review by SergeyMenshykh — APPROVED** [review]

**Review by markwallace-microsoft — APPROVED** [review]

**Pull request merged**